src/bsoncxx/test/bson_builder.cpp

```diff
@@ -1234,4 +1246,447 @@ TEST_CASE("builder::stream::document throws on consecutive keys", "[bsoncxx::bui
     REQUIRE_NOTHROW(doc << "far");
     REQUIRE_THROWS_AS(doc << "boo", bsoncxx::exception);
 }
+
+TEST_CASE("bson builder appends utf8", "[bsoncxx::builder::bson]") {
+    bson_t expected;
+    bson_init(&expected);
+
+    bson_append_utf8(&expected, "hello", -1, "world", -1);
+
+    SECTION("works with string literals") {
+        bson b{"hello", "world"};
+
+        bson_eq_document(&expected, b.view().get_document());
+    }
+
+    SECTION("works with std::string") {
+        bson b{"hello", std::string{"world"}};
+
+        bson_eq_document(&expected, b.view().get_document());
+    }
+
+    SECTION("works with b_utf8") {
+        bson b{"hello", types::b_utf8{"world"}};
+
+        bson_eq_document(&expected, b.view().get_document());
+    }
+
+    SECTION("works with const char*") {
+        const char* world = "world";
+        bson b{"hello", world};
+
+        bson_eq_document(&expected, b.view().get_document());
+    }
+
+    SECTION("works with char*") {
+        char* world = const_cast<char*>("world");
+        bson b{"hello", world};
+
+        bson_eq_document(&expected, b.view().get_document());
+    }
+
+    bson_destroy(&expected);
+}
+
+TEST_CASE("bson builder appends double", "[bsoncxx::builder::bson]") {
+    bson_t expected;
+    bson_init(&expected);
+
+    bson_append_double(&expected, "foo", -1, 1.1);
+
+    SECTION("works with raw float") {
+        bson b{"foo", 1.1};
+
+        bson_eq_document(&expected, b.view().get_document());
+    }
+
+    SECTION("works with b_double") {
+        bson b{"foo", types::b_double{1.1}};
+
+        bson_eq_document(&expected, b.view().get_document());
+    }
+
+    bson_destroy(&expected);
+}
+
+TEST_CASE("bson builder appends binary", "[bsoncxx::builder::bson]") {
+    bson_t expected;
+    bson_init(&expected);
+
+    bson_append_binary(&expected, "foo", -1, BSON_SUBTYPE_BINARY, (uint8_t*)"deadbeef", 8);
+
+    bson b{"foo", types::b_binary{binary_sub_type::k_binary, 8, (uint8_t*)"deadbeef"}};
```

Comment:
I was wondering about memory management here, with b_types that don't own their memory, but it looks like the core builder calls into the bson_append methods directly, and they copy the memory before returning from the function call, so that's fine.

Reply:
Yeah, the first step here is to implicitly construct a list builder object from all of the raw values above, e.g. `b_binary`. Since a list builder object is just a wrapper around a BSON value I generally just look at the BSON value constructors to see how these values are handled. All of the `b_types` are copied to owning BSON values before we do anything interesting with them. 

I would like to create a follow-up ticket for performance though. I haven't run any performance tests for this.